### ISSUE TITLE
Suppress hidden state output of RNNs?

### ISSUE THREAD

**Issue by MaigoAkisame** [pytorch/pytorch#3931]
Currently the forward method of RNN modules (LSTM, GRU, etc) return the output as well as the hidden states of all layers. When an RNN module contains many layers, the second output can take up a significant amount of memory, even though in many cases it is not needed. E.g.
```
lstm = nn.LSTM(40, 320, 5)
y, _ = lstm(x)    # The '_' variable takes lots of memory even though it's discarded right away
```
On the other hand, if I break down a multilayer RNN module into many modules representing single layers, I can discard the second output of the forward method layer by layer. With a 5-layer LSTM, I found doing so saved so much memory that I could double the batch size.
```
lstm1 = nn.LSTM(40, 320, 1)
lstm2 = nn.LSTM(320, 320, 1)
lstm3 = nn.LSTM(320, 320, 1)
lstm4 = nn.LSTM(320, 320, 1)
lstm5 = nn.LSTM(320, 320, 1)
x, _ = lstm1(x)    # The unwanted second output is discarded layer by layer
x, _ = lstm2(x)    # and doesn't not consume too much memory
x, _ = lstm3(x)
x, _ = lstm4(x)
x, _ = lstm5(x)
```
Is it possible to add a switch to the forward method to specify whether the second output should be computed? E.g.
```
lstm = nn.LSTM(40, 320, 5)
y = lstm(x, requires_hidden = False)    # Do not store the hidden states of each layer
```

**Comment by apaszke** [comment]
Is it seriously such a big difference in your case? Hidden states are fairly small, and we return them only for the last time step.

**Comment by MaigoAkisame** [comment]
Oh I didn't know the hidden states were only for the last time step.
But I did observe a huge difference in the batch size I could use.
Let me try if I can reproduce the problem with a smaller fragment of code.

**Comment by MaigoAkisame** [comment]
@apaszke I have made a minimum example that causes the out-of-memory error.

The code generates eight batches of random data. Their shapes are identical to the actual data I used in my experiment. Note that the batches have different shapes while having similar sizes.  This is because the sequences in different batches have different lengths. This is the key to generating the out-of-memory error.

The code implements a multi-layer LSTM in two ways: Net1 uses a single LSTM module for all the layers; Net2 uses one LSTM module for each layer.
If we use Net1, an out-of-memory error will occur on the last batch. If we use Net2, there's no problem.

I ran the code on a Tesla K20 GPU, which has around 4.7G memory.
I'm not sure if the error can be reproduced on other GPUs, but let me know if you have difficulty.

```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.autograd import Variable

class Net1(nn.Module):
    def __init__(self, nInput, nHidden, nLayers, nOutput):
        super(Net1, self).__init__()
        self.lstm = nn.LSTM(nInput, nHidden, nLayers, bidirectional = True)
        self.fc = nn.Linear(nHidden * 2, nOutput)

    def forward(self, x):
        x = self.lstm(x)[0]
        return F.log_softmax(self.fc(x))

class Net2(nn.Module):
    def __init__(self, nInput, nHidden, nLayers, nOutput):
        super(Net2, self).__init__()
        self.nLayers = nLayers
        self.lstm0 = nn.LSTM(nInput, nHidden, 1, bidirectional = True)
        for i in range(1, nLayers):
            self.__setattr__('lstm%d' % i, nn.LSTM(nHidden * 2, nHidden, 1, bidirectional = True))
        self.fc = nn.Linear(nHidden * 2, nOutput)

    def forward(self, x):
        for i in range(self.nLayers):
            x = self.__getattr__('lstm%d' % i)(x)[0]
        return F.log_softmax(self.fc(x))

model = Net1(nInput = 40, nHidden = 320, nLayers = 5, nOutput = 40).cuda()

shapes = [(1469, 13, 40), (1429, 13, 40), (1414, 14, 40),
          (1395, 14, 40), (1374, 14, 40), (1358, 14, 40),
          (1346, 14, 40), (1329, 15, 40)]
cv_x = [Variable(torch.rand(shape)) for shape in shapes]

for i in range(len(cv_x)):
    print i
    pred = model(cv_x[i].cuda())
```

**Comment by apaszke** [comment]
Thank you! We'll look into that.

**Comment by MaigoAkisame** [comment]
I just found that my experiment using Net2 crashed at the beginning of the second epoch because of Out of Memory. Between the first and second epochs it evaluated the performance on the validation set, which might have consumed some memory so that the second epoch couldn't run.

The RNN module seems really thirsty on memory, or maybe I did something wrong...

**Comment by gailweiss** [comment]
similarly - could we have an option to return just the very last state? Getting all of them on long sequences when I just want the last - e.g. when preparing for truncated backprop - blows up the memory for no reason
